docs/guides/go-built-in.md

```diff
@@ -311,112 +311,90 @@ information on why the transaction was rejected.
-confuse them with the transactions included in the blocks delivered by CometBFT,
-the _application transactions_.
-
-First, let's create a new Badger transaction during `BeginBlock`. All application transactions in the
-current block will be executed within this Badger transaction.
-Then, return informing CometBFT that the application is ready to receive application transactions:
+The `FinalizeBlock` method executes the block, including any necessary transaction processing and state updates, and returns a `ResponseFinalizeBlock` object which contains any necessary information about the executed block.
 
 ```go
-func (app *KVStoreApplication) BeginBlock(req abcitypes.RequestBeginBlock) abcitypes.ResponseBeginBlock {
-	app.onGoingBlock = app.db.NewTransaction(true)
-	return abcitypes.ResponseBeginBlock{}
-}
-```
-
-Next, let's modify `DeliverTx` to add the `key` and `value` to the database transaction every time our application
-receives a new application transaction through `RequestDeliverTx`.
-
-```go
-func (app *KVStoreApplication) DeliverTx(req abcitypes.RequestDeliverTx) abcitypes.ResponseDeliverTx {
-	if code := app.isValid(req.Tx); code != 0 {
-		return abcitypes.ResponseDeliverTx{Code: code}
-	}
-
-	parts := bytes.SplitN(req.Tx, []byte("="), 2)
-	key, value := parts[0], parts[1]
-
-	if err := app.onGoingBlock.Set(key, value); err != nil {
-		log.Panicf("Error writing to database, unable to execute tx: %v", err)
-	}
-
-	return abcitypes.ResponseDeliverTx{Code: 0}
+func (app *KVStoreApplication) FinalizeBlock(_ context.Context, req *abcitypes.RequestFinalizeBlock) (*abcitypes.ResponseFinalizeBlock, error) {
+    var txs = make([]*abcitypes.ExecTxResult, len(req.Txs))
+
+    app.onGoingBlock = app.db.NewTransaction(true)
+    for i, tx := range req.Txs {
+        if code := app.isValid(tx); code != 0 {
+            log.Printf("Error in tx in if")
+            txs[i] = &abcitypes.ExecTxResult{Code: code}
+        } else {
+            parts := bytes.SplitN(tx, []byte("="), 2)
+            key, value := parts[0], parts[1]
+            log.Printf("Adding key %s with value %s", key, value)
+
+            if err := app.onGoingBlock.Set(key, value); err != nil {
+                log.Panicf("Error writing to database, unable to execute tx: %v", err)
+            }
+
+            log.Printf("Successfully added key %s with value %s", key, value)
+
+            // Marshal the ExecTxResult to the txs slice element
+            txs[i] = &abcitypes.ExecTxResult{}
+            if _, err := txs[i].MarshalToSizedBuffer(nil); err != nil {
+                log.Panicf("Error marshalling tx result: %v", err)
+            }
+        }
+    }
+
+    return &abcitypes.ResponseFinalizeBlock{
+    TxResults:        txs,
+    AppHash:          []byte{},
+    Events:           []abcitypes.Event{},
+    ValidatorUpdates: abcitypes.ValidatorUpdates{},
+    }, nil
 }
 ```
 
-Note that we check the validity of the transaction _again_ during `DeliverTx`.
```

Comment:
I think this is important. It could be captured in a best practices document or left here.